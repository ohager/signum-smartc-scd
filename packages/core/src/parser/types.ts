export type DataType = Array<DataType> | String | Boolean;
export type MapType = {
  name: string;
  description?: string;
  type?: string;
  constant?: boolean;
  value?: DataType;
  oneOf?: Array<{
    name: string;
    value: string;
  }>;
};

export interface VariableDefinition {
  name: string;
  type: DataType;
  description?: string;
  initializable?: boolean;
  constant?: boolean;
  // for structs only
  fields?: Array<{
    name: string;
    type: DataType;
  }>;
}

export interface MapDefinition {
  name: string;
  key1: MapType;
  key2: MapType;
  value: MapType;
}

export interface TransactionDefinition {
  name: string;
  kind:
    | "sendAmountAndMessage"
    | "sendBalance"
    | "sendQuantity"
    | "sendQuantityAndAmount";
  inputs: Array<{
    name: string;
    type: DataType;
  }>;
}

export interface FunctionDefinition {
  name: string;
  description?: string;
  code: string;
  args: {
    name: string;
    type: DataType;
  }[];
}

export interface ABIType {
  contractName: string;
  description?: string;
  activationAmount: string;
  pragmas: {
    maxAuxVars: number;
    verboseAssembly: boolean;
    optimizationLevel?: number;
    version: string;
    codeStackPages?: number;
    userStackPages?: number;
  };
  functions: FunctionDefinition[];
  stateLayout: VariableDefinition[];
  maps: MapDefinition[];
  transactions: TransactionDefinition[];
}
