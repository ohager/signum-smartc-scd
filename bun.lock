{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "signum-smartc-abi",
      "devDependencies": {
        "@types/bun": "latest",
        "turbo": "^2.5.0",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
    "packages/core": {
      "name": "@signum-smartc-abi/core",
      "dependencies": {
        "ajv": "^8.17.1",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@signum-smartc-abi/core": ["@signum-smartc-abi/core@workspace:packages/core"],

    "@types/bun": ["@types/bun@1.2.9", "", { "dependencies": { "bun-types": "1.2.9" } }, "sha512-epShhLGQYc4Bv/aceHbmBhOz1XgUnuTZgcxjxk+WXwNyDXavv5QHD1QEFV0FwbTSQtNq6g4ZcV6y0vZakTjswg=="],

    "@types/node": ["@types/node@22.14.1", "", { "dependencies": { "undici-types": "~6.21.0" } }, "sha512-u0HuPQwe/dHrItgHHpmw3N2fYCR6x4ivMNbPHRkBVP4CvN+kiRrKHWk3i8tXiO/joPwXLMYvF9TTF0eqgHIuOw=="],

    "@types/ws": ["@types/ws@8.18.1", "", { "dependencies": { "@types/node": "*" } }, "sha512-ThVF6DCVhA8kUGy+aazFQ4kXQ7E1Ty7A3ypFOe0IcJV8O/M511G99AW24irKrW56Wt44yG9+ij8FaqoBGkuBXg=="],

    "ajv": ["ajv@8.17.1", "", { "dependencies": { "fast-deep-equal": "^3.1.3", "fast-uri": "^3.0.1", "json-schema-traverse": "^1.0.0", "require-from-string": "^2.0.2" } }, "sha512-B/gBuNg5SiMTrPkC+A2+cW0RszwxYmn6VYxB/inlBStS5nx6xHIt/ehKRhIMhqusl7a8LjQoZnjCs5vhwxOQ1g=="],

    "bun-types": ["bun-types@1.2.9", "", { "dependencies": { "@types/node": "*", "@types/ws": "*" } }, "sha512-dk/kOEfQbajENN/D6FyiSgOKEuUi9PWfqKQJEgwKrCMWbjS/S6tEXp178mWvWAcUSYm9ArDlWHZKO3T/4cLXiw=="],

    "fast-deep-equal": ["fast-deep-equal@3.1.3", "", {}, "sha512-f3qQ9oQy9j2AhBe/H9VC91wLmKBCCU/gDOnKNAYG5hswO7BLKj09Hc5HYNz9cGI++xlpDCIgDaitVs03ATR84Q=="],

    "fast-uri": ["fast-uri@3.0.6", "", {}, "sha512-Atfo14OibSv5wAp4VWNsFYE1AchQRTv9cBGWET4pZWHzYshFSS9NQI6I57rdKn9croWVMbYFbLhJ+yJvmZIIHw=="],

    "json-schema-traverse": ["json-schema-traverse@1.0.0", "", {}, "sha512-NM8/P9n3XjXhIZn1lLhkFaACTOURQXjWhV4BA/RnOv8xvgqtqpAX9IO4mRQxSx1Rlo4tqzeqb0sOlruaOy3dug=="],

    "require-from-string": ["require-from-string@2.0.2", "", {}, "sha512-Xf0nWe6RseziFMu+Ap9biiUbmplq6S9/p+7w7YXP/JBHhrUDDUhwa+vANyubuqfZWTveU//DYVGsDG7RKL/vEw=="],

    "turbo": ["turbo@2.5.0", "", { "optionalDependencies": { "turbo-darwin-64": "2.5.0", "turbo-darwin-arm64": "2.5.0", "turbo-linux-64": "2.5.0", "turbo-linux-arm64": "2.5.0", "turbo-windows-64": "2.5.0", "turbo-windows-arm64": "2.5.0" }, "bin": { "turbo": "bin/turbo" } }, "sha512-PvSRruOsitjy6qdqwIIyolv99+fEn57gP6gn4zhsHTEcCYgXPhv6BAxzAjleS8XKpo+Y582vTTA9nuqYDmbRuA=="],

    "turbo-darwin-64": ["turbo-darwin-64@2.5.0", "", { "os": "darwin", "cpu": "x64" }, "sha512-fP1hhI9zY8hv0idym3hAaXdPi80TLovmGmgZFocVAykFtOxF+GlfIgM/l4iLAV9ObIO4SUXPVWHeBZQQ+Hpjag=="],

    "turbo-darwin-arm64": ["turbo-darwin-arm64@2.5.0", "", { "os": "darwin", "cpu": "arm64" }, "sha512-p9sYq7kXH7qeJwIQE86cOWv/xNqvow846l6c/qWc26Ib1ci5W7V0sI5thsrP3eH+VA0d+SHalTKg5SQXgNQBWA=="],

    "turbo-linux-64": ["turbo-linux-64@2.5.0", "", { "os": "linux", "cpu": "x64" }, "sha512-1iEln2GWiF3iPPPS1HQJT6ZCFXynJPd89gs9SkggH2EJsj3eRUSVMmMC8y6d7bBbhBFsiGGazwFIYrI12zs6uQ=="],

    "turbo-linux-arm64": ["turbo-linux-arm64@2.5.0", "", { "os": "linux", "cpu": "arm64" }, "sha512-bKBcbvuQHmsX116KcxHJuAcppiiBOfivOObh2O5aXNER6mce7YDDQJy00xQQNp1DhEfcSV2uOsvb3O3nN2cbcA=="],

    "turbo-windows-64": ["turbo-windows-64@2.5.0", "", { "os": "win32", "cpu": "x64" }, "sha512-9BCo8oQ7BO7J0K913Czbc3tw8QwLqn2nTe4E47k6aVYkM12ASTScweXPTuaPFP5iYXAT6z5Dsniw704Ixa5eGg=="],

    "turbo-windows-arm64": ["turbo-windows-arm64@2.5.0", "", { "os": "win32", "cpu": "arm64" }, "sha512-OUHCV+ueXa3UzfZ4co/ueIHgeq9B2K48pZwIxKSm5VaLVuv8M13MhM7unukW09g++dpdrrE1w4IOVgxKZ0/exg=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "undici-types": ["undici-types@6.21.0", "", {}, "sha512-iwDZqg0QAGrg9Rav5H4n0M64c3mkR59cJ6wQp+7C4nI0gsmExaedaYLNO44eT4AtBBwjbTiGPMlt2Md0T9H9JQ=="],
  }
}
