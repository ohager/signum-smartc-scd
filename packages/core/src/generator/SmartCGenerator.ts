import { Eta } from "eta";
import type { SCD } from "../parser";
import {SmartCTemplate} from "./templates/smartc.eta"

export class SmartCGenerator {
  private eta: Eta;

  constructor(private scd: SCD) {
    this.eta = new Eta();
  }

  generateContract(): string {
    const templateData = {
      contractName: this.scd.getContractInfo().name,
      description: this.scd.getContractInfo().description,
      activationAmount: this.scd.getContractInfo().activationAmount,
      pragmas: this.scd.getContractInfo().pragmas,
      methods: this.scd.getMethods(),
      variables: this.scd.getVariables(),
      structs: this.scd.getStructs(),
      maps: this.scd.getMaps(),
    };

    return this.eta.renderString(SmartCTemplate, templateData);
  }
}
