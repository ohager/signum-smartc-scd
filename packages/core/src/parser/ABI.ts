import Ajv from "ajv";
import schema from "./abi-schema.json";
import type {
  FunctionDefinition,
  ABIType,
  MapDefinition,
  VariableDefinition,
  TransactionDefinition,
} from "./types";

/**
 * This is the ABI class which provides convenience functions for further tooling.
 */
export class ABI {
  private static ajv = new Ajv();
  private static validate = ABI.ajv.compile<ABIType>(schema);

  private constructor(private abi: ABIType) {}

  static parse(input: string | object) {
    const valid = this.validate(input);

    if (!valid) {
      const errors = this.validate.errors;
      throw new Error(`Invalid ABI: ${JSON.stringify(errors)}`);
    }
    return new ABI(input);
  }

  getStateLayout() {
    let currentIndex = this.abi.pragmas.maxAuxVars + 1;
    const layout: (VariableDefinition & { index: number })[] = [];
    for (const variable of this.abi.stateLayout) {
      if (variable.type === "struct") {
        // Handle struct fields
        const fields =
          variable.fields?.map((field) => ({
            ...field,
            name: `${variable.name}.${field.name}`,
            index: currentIndex++,
          })) || [];

        layout.push(...fields);
      } else {
        layout.push({
          ...variable,
          index: currentIndex++,
        });
      }
    }

    return layout;
  }

  getFunctions(): FunctionDefinition[] {
    return this.abi.functions.map((f) => ({
      ...f,
      code: BigInt(f.code),
    }));
  }

  getMaps(): MapDefinition[] {
    return this.abi.maps;
  }

  getStateVariables(): VariableDefinition[] {
    return this.abi.stateLayout.filter(
      (variable) => variable.type !== "struct",
    );
  }

  getStructs(): VariableDefinition[] {
    return this.abi.stateLayout.filter(
      (variable) => variable.type === "struct",
    );
  }

  getTransactions(): TransactionDefinition[] {
    return this.abi.transactions;
  }

  getContractInfo() {
    return {
      name: this.abi.contractName,
      description: this.abi.description,
      activationAmount: BigInt(this.abi.activationAmount),
      pragmas: this.abi.pragmas,
    };
  }
}
