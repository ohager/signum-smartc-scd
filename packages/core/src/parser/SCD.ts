import Ajv from "ajv";
import schema from "./scd-schema.json";
import type {
  MethodDefinition,
  SCDType,
  MapDefinition,
  VariableDefinition,
  TransactionDefinition,
} from "./types";

/**
 * This is the ABI class which provides convenience functions for further tooling.
 */
export class SCD {
  private static ajv = new Ajv();
  private static validate = SCD.ajv.compile<SCDType>(schema);

  private constructor(private scd: SCDType) {}

  static parse(input: string | object) {
    const valid = this.validate(input);

    if (!valid) {
      const errors = this.validate.errors;
      throw new Error(`Invalid SCD: ${JSON.stringify(errors)}`);
    }
    return new SCD(input);
  }

  get raw(): SCDType { return this.scd}

  getVariablesLayout() {
    let currentIndex = this.scd.pragmas.maxAuxVars + 1;
    const layout: (VariableDefinition & { index: number })[] = [];
    for (const variable of this.scd.variables) {
      if (variable.type === "struct") {
        // Handle struct fields
        const fields =
          variable.fields?.map((field) => ({
            ...field,
            name: `${variable.name}.${field.name}`,
            index: currentIndex++,
          })) || [];

        layout.push(...fields);
      } else {
        layout.push({
          ...variable,
          index: currentIndex++,
        });
      }
    }

    return layout;
  }

  getMethods(): Readonly<MethodDefinition[]> {
    return this.scd.methods;
  }

  getMaps(): Readonly<MapDefinition[]> {
    return this.scd.maps;
  }

  getVariables(): Readonly<VariableDefinition[]> {
    return this.scd.variables.filter((variable) => variable.type !== "struct");
  }

  getStructs(): Readonly<VariableDefinition[]> {
    return this.scd.variables.filter((variable) => variable.type === "struct");
  }

  getTransactions(): Readonly<TransactionDefinition[]> {
    return this.scd.transactions;
  }

  getContractInfo() {
    return {
      name: this.scd.contractName,
      description: this.scd.description,
      activationAmount: this.scd.activationAmount,
      pragmas: this.scd.pragmas,
    };
  }
}
