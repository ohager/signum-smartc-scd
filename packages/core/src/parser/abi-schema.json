{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "SmartC ABI Schema",
  "definitions": {
    "DataType": {
      "type": "string",
      "enum": [
        "address",
        "string",
        "boolean",
        "long",
        "amount",
        "txId",
        "address[]",
        "string[]",
        "boolean[]",
        "long[]",
        "amount[]",
        "txId[]",
        "struct"
      ]
    },
    "MapType": {
      "type": "object",
      "required": ["name"],
      "properties": {
        "name": { "type": "string" },
        "description": { "type": "string" },
        "constant": { "type": "boolean", "default": false },
        "type": { "$ref": "#/definitions/DataType" },
        "value": { "type": "string" },
        "oneOf": {
          "type": "array",
          "items": {
            "type": "object",
            "required": ["name", "value"],
            "properties": {
              "name": { "type": "string" },
              "value": { "type": "string" }
            }
          }
        }
      }
    }
  },
  "type": "object",
  "required": [
    "contractName",
    "activationAmount",
    "pragmas",
    "methods",
    "stateLayout",
    "maps"
  ],
  "properties": {
    "contractName": {
      "type": "string",
      "description": "Name of the smart contract"
    },
    "description": {
      "type": "string",
      "description": "Description of the contract's purpose"
    },
    "activationAmount": {
      "type": "string",
      "pattern": "^[0-9_]+$",
      "description": "Amount required to activate the contract"
    },
    "codeStackPages": { "type": "number", "minimum": 0, "maximum": 10 },
    "userStackPages": { "type": "number", "minimum": 0, "maximum": 10 },
    "pragmas": {
      "type": "object",
      "properties": {
        "maxAuxVars": { "type": "number", "minimum": 0, "maximum": 10 },
        "maxConstVars": { "type": "number", "minimum": 0, "maximum": 10 },
        "optimizationLevel": { "type": "number", "minimum": 0, "maximum": 4 },
        "version": { "type": "string" },
        "reuseAssignedVar": { "type": "boolean" },
        "verboseAssembly": { "type": "boolean" },
        "verboseScope": { "type": "boolean" }
      }
    },
    "methods": {
      "type": "array",
      "items": {
        "type": "object",
        "required": ["name", "code", "args"],
        "properties": {
          "name": { "type": "string" },
          "description": { "type": "string" },
          "code": { "type": "string" },
          "args": {
            "type": "array",
            "items": {
              "type": "object",
              "required": ["name", "type"],
              "properties": {
                "name": { "type": "string" },
                "type": { "$ref": "#/definitions/DataType" }
              }
            }
          }
        }
      }
    },
    "stateLayout": {
      "type": "array",
      "items": {
        "type": "object",
        "required": ["name", "type"],
        "properties": {
          "name": { "type": "string" },
          "type": {
            "type": "string",
            "enum": [
              "address",
              "string",
              "boolean",
              "amount",
              "number",
              "struct",
              "array"
            ]
          },
          "description": { "type": "string" },
          "initializable": { "type": "boolean" },
          "constant": { "type": "boolean" },
          "fields": {
            "type": "array",
            "items": {
              "type": "object",
              "required": ["name", "type"],
              "properties": {
                "name": { "type": "string" },
                "type": { "type": "string" }
              }
            }
          }
        }
      }
    },
    "maps": {
      "type": "array",
      "items": {
        "type": "object",
        "required": ["name", "key1", "key2", "value"],
        "properties": {
          "name": { "type": "string" },
          "key1": {
            "$ref": "#/definitions/MapType"
          },
          "key2": {
            "$ref": "#/definitions/MapType"
          },
          "value": {
            "$ref": "#/definitions/MapType"
          }
        }
      }
    },
    "transactions": {
      "type": "array",
      "items": {
        "type": "object",
        "required": ["name", "kind", "inputs"],
        "properties": {
          "name": { "type": "string" },
          "kind": {
            "type": "string",
            "enum": [
              "sendAmountAndMessage",
              "sendBalance",
              "sendQuantity",
              "sendQuantityAndAmount"
            ]
          },
          "inputs": {
            "type": "array",
            "items": {
              "type": "object",
              "required": ["name", "type"],
              "properties": {
                "name": { "type": "string" },
                "type": { "$ref": "#/definitions/DataType" }
              }
            }
          }
        }
      }
    }
  }
}
