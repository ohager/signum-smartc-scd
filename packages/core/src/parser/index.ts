export {ABI} from  "./ABI"
export * from "./types"
