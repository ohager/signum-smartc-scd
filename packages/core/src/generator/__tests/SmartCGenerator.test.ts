import { SCD } from "../../parser";
import { SmartCGenerator } from "../SmartCGenerator";
import { describe, expect, it } from "bun:test";

const mockSCD = {
  contractName: "TestContract",
  contractHash: "0x1234567890abcdef",
  description: "A test contract",
  activationAmount: "1000000",
  pragmas: {
    maxAuxVars: 3,
    verboseAssembly: true,
    optimizationLevel: 3,
    version: "2.2.1",
    codeStackPages: 0,
    userStackPages: 0,
  },
  methods: [
    {
      name: "testMethod",
      code: "100",
      args: [
        { name: "param1", type: "long" },
        { name: "param2", type: "address" },
      ],
    },
  ],
  variables: [
    {
      name: "owner",
      type: "address",
      initializable: true,
      constant: true,
    },
    {
      name: "stats",
      type: "struct",
      fields: [
        { name: "counter", type: "long" },
        { name: "balance", type: "amount" },
      ],
    },
  ],
  maps: [
    {
      name: "testMap",
      key1: {
        name: "key1",
        type: "long",
        description: "Main Key",
        constant: true,
        value: "1",
      },
      key2: {
        name: "addr",
        description: "Address",
        type: "address",
        constant: false,
      },
      value: {
        name: "value",
        description: "Value",
        constant: false,
        oneOf: [
          {
            name: "SOME_CONSTANT_1",
            value: "1",
          },
          {
            name: "SOME_CONSTANT_2",
            value: "2",
          },
        ],
      },
    },
  ],
};

describe("SmartCGenerator", () => {
  it("should generate SmartC code", async () => {
    const generator = new SmartCGenerator(SCD.parse(mockSCD));
    const smartCCode = await generator.generateContract();
    expect(smartCCode).toMatchSnapshot();
  });
});
