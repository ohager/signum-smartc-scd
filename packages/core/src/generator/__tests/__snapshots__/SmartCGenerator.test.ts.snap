// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartCGenerator should generate SmartC code 1`] = `
"#program name TestContract
#program description A test contract
#program activationAmount 1000000
#pragma maxAuxVars 3
#pragma verboseAssembly true
#pragma optimizationLevel 3
#pragma version 2.3.0
#pragma codeStackPages 0
#pragma userStackPages 0

// Magic codes for methods
#define TESTMETHOD 100
#define TESTMETHOD2 101

// Maps
#define MAP_TESTMAP_KEY1 1


// State variables
long owner;


// Structs
struct STATS {
        long counter;
        long balance;
    } stats;


void main() {
    while ((currentTx.txId = getNextTx()) != 0) {
        currentTx.sender = getSender(currentTx.txId);
        readMessage(currentTx.txId, 0, currentTx.message);

        switch(currentTx.message[0]) {
                    case TESTMETHOD:
                testMethod(currentTx.message[1], currentTx.message[2]);
                break;
                    case TESTMETHOD2:
                testMethod2(currentTx.message[1]);
                break;
                }
    }
}



// Function stubs
void testMethod(long param1, long param2) {
    // TODO: Implement testMethod
}
void testMethod2(long param1) {
    // TODO: Implement testMethod2
}
"
`;
